truffle/src/main/java/org/jruby/truffle/runtime/rope/Rope.java

```diff
@@ -93,4 +106,34 @@ public int realSize() {
     public int getRealSize() {
         return realSize();
     }
+
+    @Override
+    public int hashCode() {
+        if (hashCode == 0) {
+            hashCode = RopeOperations.hashForRange(this, 1, 0, byteLength);
+        }
+
+        return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o instanceof Rope) {
+            final Rope other = (Rope) o;
+
+            if ((hashCode != 0) && (other.hashCode != 0) && (hashCode != other.hashCode)) {
+                return false;
+            }
+
+            // TODO (nirvdrum 21-Jan-16): We really should be taking the encoding into account here. We're currenly not because it breaks the symbol table.
```

Comment:
We were relying on `ByteList#equals` to not take the encoding into account. We currently expect two Symbols with different encodings but the same bytes to be the same Symbol. This is not what MRI.

As a simple example:

```
:abc.encoding => Encoding:US-ASCII
"abc".to_sym.encoding => Encoding:US-ASCII
"def".to_sym.encoding => Encoding:UTF-8
:def.encoding => Encoding:UTF-8
```

How a Symbol ends up in the table in the first place ends up affecting the encoding for future operations that look it up. When I first attempted to fix this, we ended up with duplicate entries (one for UTF-8, one for US-ASCII) and it broke a lot of caches. That fix needs to be combined with something else that converts UTF-8 to US-ASCII on the `String#to_sym` path. I'm not entirely clear on the semantics of maintaining encodings from Strings to Symbols.